core/templates/dev/head/collection_player/CollectionDataService.js

```diff
@@ -0,0 +1,126 @@
+// Copyright 2015 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service to retrieve information about collections from the
+ * backend.
+ *
+ * @author henning.benmax@gmail.com (Ben Henning)
+ */
+
+// TODO(bhenning): For preview mode, this service should be omitted. Instead,
+// a separate collectionDataService should be created which serves a local copy
+// of the collection instead.
+oppia.factory('CollectionDataService', [
+    '$http', 'COLLECTION_DATA_URL', 'UrlInterpolationService',
+    function($http, COLLECTION_DATA_URL, urlInterpolationService) {
+
+  // Maps previously loaded collections to their IDs.
+  var _collectionCache = [];
+
+  // This inserts an extra 'exploration' object into each node. The object
+  // includes the exploration ID and additional meta-information regarding the
+  // exploration that is useful for collection pages (such as the exploration
+  // title).
+  var _populateExplorationsInNodes = function(collection, explorationTitles) {
+    for (var i = 0; i < collection.nodes.length; i++) {
+      var collectionNode = collection.nodes[i];
+      collectionNode.exploration = {
+        'id': collectionNode.exploration_id,
+        'title': explorationTitles[collectionNode.exploration_id]
+      };
+    }
+    return collection;
+  };
+
+  // TODO(bhenning): Figure out what to do about versioning. Should the version
+  // be specified when a collection is fetched?
+  var _fetchCollection = function(
+      collectionId, successCallback, errorCallback) {
+    var collectionDataUrl = urlInterpolationService.interpolateUrl(
+      COLLECTION_DATA_URL, {'collection_id': collectionId});
+
+    $http.get(collectionDataUrl).success(function(data) {
+      var collection = _populateExplorationsInNodes(
+        angular.copy(data.collection), data.exploration_titles);
+      collection.next_exploration_ids = data.next_exploration_ids;
+      if (successCallback) {
+        successCallback(collection);
+      }
+    }).error(function(data) {
+      if (errorCallback) {
+        errorCallback(data ? data.error : null, collectionId);
```

Comment:
I don't think I see the issue. What's an example of the case you're defending against?

I'm concerned about this because it feels a bit non-standard to pass additional 'bits' of data like this into the errorCallback. I'd have expected 'data' to contain this information, or the caller to keep a note of it (I mean, errorCallback is being passed into this function in the same way that collectionId is, so I don't see why the caller can't set the collectionId that ends up getting used in errorCallback).